tiledb/sm/array/array_directory.cc

```diff
@@ -393,8 +393,37 @@ Status ArrayDirectory::load() {
           "Cannot open array; Array does not exist."));
     }
 
-    // Set the latest array schema URI
-    latest_array_schema_uri_ = array_schema_uris_.back();
+    // Set the latest array schema URI. When in READ mode, the latest array
+    // schema URI is the schema with the largest timestamp less than or equal
+    // to the current timestamp_end_. If no schema meets this definition, we
+    // use the first schema available.
+    if (mode_ == ArrayDirectoryMode::READ) {
+      optional<URI> latest_uri = nullopt;
+      uint64_t latest_ts = 0;
+
+      for (auto& uri : array_schema_uris_) {
+        auto name = uri.remove_trailing_slash().last_path_part();
+
+        // Skip the old schema URI name since it doesn't have timestamps
+        if (name == constants::array_schema_filename) {
+          continue;
+        }
```

Comment:
I believe there's a defect here that if there's a timestamp requested that precedes any of the files that includes times, it won't find anything. It should probably find this one, but it's unambiguously discarded.

Similarly, I find it odd that this code doesn't throw anywhere. If a timestamp is requested that precedes all available schema, it should fail.

Reply:
> I believe there's a defect here that if there's a timestamp requested that precedes any of the files that includes times, it won't find anything. It should probably find this one, but it's unambiguously discarded.

I'm guessing you missed this line?

https://github.com/TileDB-Inc/TileDB/pull/4549/files#diff-c2be2932fddd2739bad67ac6b693a91fe96d060d291998be7082e78057fc850dR422

> Similarly, I find it odd that this code doesn't throw anywhere. If a timestamp is requested that precedes all available schema, it should fail.

Yeah, I started off thinking that as well but then I got to considering all of the different edge cases around our requirements for reading arrays that have a `__array_schema.tdb` and I realized we can't blindly fail here since those arrays have an un-timestamped schema available.

The other thing to keep in mind is that we can time travel before the first schema was written and write fragments. This is done a lot in the test suite, but more importantly, its possible users have done this. So changing the array schema latest could theoretically break some users existing arrays if they're doing fun things like that.

Reply:
>we can time travel before the first schema was written and write fragments

It's a defect that we allow this. Such data is fundamentally ambiguous, because it could have been written with any schema. Rather than double down on supporting defective behavior, I'd prefer that we fail to read such data with a "contact TileDB" message. We might be able to infer what the data means from the schema history, but that's a manual process.

Reply:
@eric-hughes-tiledb this is out of scope for 2.19 and we need this PR in for the release per discussion with Isaiah. If we do what you suggest we are going to end up with a lot of support calls from customers as we have a lot of arrays out there that do this. Let’s file a follow up story for this an take more time to figure out how we can correct this moving forward.